python/simulation.cpp

```diff
@@ -210,7 +212,7 @@ void register_simulation(pybind11::module& m, pyarb_global_ptr global_ptr) {
                               const std::optional<arb::domain_decomposition>& decomp,
                               std::uint64_t seed) {
                 try {
-                    auto ctx = ctx_ ? ctx_ : std::make_shared<context_shim>(arb::make_context());
+                    auto ctx = ctx_ ? ctx_ : std::make_shared<context_shim>(arb::make_context({arbenv::thread_concurrency(),-1}));
```

Comment:
So, under the idea that Python and C++ APIs should co-evolve I am against this. Keep it at the default here and change the default in C++. But still keep the option to ask for 'avail_threads' explicitly.

Reply:
So, also in cpp make `avail_threads` an option you mean?

Reply:
Well, currently it is, right, via `thread_concurrency`?

Reply:
Ah, sure, but it isn't a default (yet). I thought that was your request?

I was thinking that the best place to set this default (once) is in `proc_allocation`, but no, because it does not know about MPI. So, the default `context` constructors should set `num_threads` to `arbenv::thread_concurrency()`, if MPI is not enabled.

In the interest of keeping things similar, should in Python the variable be called `thread_concurrency`, and be a member of `arbor`? e.g. `arbor.context(arbor.thread_concurrency(), None)`? I like it!

Reply:
Should be in `arbenv`, yes, but that's the vision.

Reply:
Unlike in Python, there is no explicit constructor in C++, it all goes through `proc_allocation`: `execution_context(const proc_allocation& resources = proc_allocation{})`. Also, having this in C++ would make (all of) Arbor dependent on `arbenv`.

Since there is no way to currently do sth like `auto ctx =  arbor::context{gpu_id=1}`, we could set `proc_allocation.num_threads` by changing the non-mpi constructor to `execution_context(const proc_allocation& resources = proc_allocation{arbenv::thread_concurrency(),-1});` To the extent that `execution_context` construction maps to Python, it sets `num_threads` under the same circumstances.

Reply:
> Unlike in Python, there is no explicit constructor in C++, it all goes through proc_allocation: execution_context(const proc_allocation& resources = proc_allocation{}). 

You could call proc_allocation a set of kwargs. Once we have achieved C++20 in its full glory this becomes 
a thing
```cxx
execution_context({.threads=4, .comm=MPI_COMM_WORLD})
```

> Also, having this in C++ would make (all of) Arbor dependent on arbenv.

That's fine, to a certain extent, `arbenv` is a library made for Arbor to interact with the environment.
As long we do not pull it into the deepest layers.

Also, nothing stopping us from making new, purpose-built c'tors, eg
```cxx
context from_mpi_one_thread_per_logical_core(MPI_Comm& comm, int gpu_id=-1);
context from_mpi_one_thread_per_physical_core(MPI_Comm& comm, int gpu_id=-1);
// ...
```

Also, if proc_allocation is really too much of a hassle to have around, feel free to remove it.

Reply:
I've updated the code as per my earlier suggestion in the interested of merging this now.

- Removal of proc_allocation is fine by me, but I'm not a user. The C++ tests do, so that's why I have not suggested removal.
- MPI cleverness: to me that can be seen as a separate issues (the PR was initially written to get a better default for local use). I'll ask in the channel if users are missing something.